spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala

```diff
@@ -425,14 +431,70 @@ class DLModel[@specialized(Float, Double) T: ClassTag](
   }
 
   override def copy(extra: ParamMap): DLModel[T] = {
-    val copied = new DLModel(model, featureSize, uid).setParent(parent)
+    val copied = new DLModel(_model, featureSize, uid).setParent(parent)
     copyValues(copied, extra)
   }
+
+  private var _model: Module[T] = model
+
+  override protected def getModule: Module[T] = _model
+
+  override protected def setModule(module: Module[T]): Unit = _model = module
 }
 
-// TODO, add save/load
 object DLModel {
+  def load[@specialized(Float, Double) T: ClassTag]
+  (path: String)(implicit ev: TensorNumeric[T]): DLModel[T] = {
+    DLModelBase.load[T, DLModel[T]](path).asInstanceOf[DLModel[T]]
+  }
+}
+
+/**
+ * Base class of DLModels, such as DLClassifierModel, provide (pipeline) model IO features.
+ * Please make sure working with module which returned by [[getModule]] and set by [[setModule]].
+ */
+abstract class DLModelBase[@specialized(Float, Double) T: ClassTag, M <: DLTransformerBase[M]]
+()(implicit ev: TensorNumeric[T]) extends DLTransformerBase[M] {
+  protected def getModule: Module[T]
 
+  protected def setModule(module: Module[T]): Unit
 
+  /**
+   * Save a [[DLModelBase]] to a dictionary, which contains 3 files(meta, module, weight).
+   * @param path path of model dictionary, it should NOT exists as a File.
+   * @param isOverWrite whether overwrite earlier files
+   */
+  def save(path: String, isOverWrite: Boolean): Unit = {
+    val (metaPath, modulePath, weightPath) = DLModelBase.getPaths(path)
+    val meta = SerializationUtils.serialize(this)
```

Comment:
Is it based on Java serialization? I'm a little concerned since it may not be compatible across different Java/scala version.

Reply:
Yes, it is based on Java Serialization. Spark use Java Serialization(kyro is another option, but java is default) to do all serialization works. And BigDL Module serialized by [[SerializationUtils.serialize]] at DistriOptimizer.initThreadModels in eariler version.

Reply:
ML Model and Estimator does not use Java serialization for save/load. And I would recommend we follow the implementation in ML and leverage the existing infrastructure code.

Reply:
Okay,  so,  is it a good way to do saving/loading  by extending MLReader/MLWriter :)

Reply:
Another reminder, We need to support both Spark 1.6 and Spark 2.X.